### PR TITLE
databases homework week3

### PR DESCRIPTION
**Pull request by panivita**

### CODE REVIEW

File: databases/week3/meal_sharing_db.sql [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+CREATE TABLE  `meal` (
+ `id` int(10) unsigned NOT NULL AUTO_INCREMENT PRIMARY KEY,
+ `title` VARCHAR(255) NOT NULL,
+ `description` TEXT NOT NULL,
+ `location` VARCHAR(255) NOT NULL,
+ `meal_time` DATETIME NOT NULL,
+ `max_reservations` int(10) unsigned NOT NULL,
+ `price` DECIMAL(6, 2) NOT NULL,
+ `created_date` DATE NOT NULL DEFAULT NOW()
```
Comment (flemeur) [inline databases/week3/meal_sharing_db.sql:14]: This line gives me the following error: `Error in query (1067): Invalid default value for 'created_date'`

I wasn't aware of this but MySQL apparently only supports default values for DATE columns when the default is enclosed in parenthesis. See: https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html#data-types-defaults-explicit

So you would have to write something like this:
```suggestion
 `created_date` DATE NOT NULL DEFAULT (NOW())
```

Same goes for the other DATE columns of course

File: databases/week3/meal_sharing_db.sql [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+ `description` TEXT NOT NULL,
+ `meal_id` int(10) unsigned NOT NULL,
+ `stars` int(10) unsigned NOT NULL,
+ `created_date` DATE NOT NULL DEFAULT NOW(),
+ CONSTRAINT `fk_review_meal` FOREIGN KEY (`meal_id`) REFERENCES `meal` (`id`) ON DELETE CASCADE ON UPDATE CASCADE
+ ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
+ 
+ -- Meal
+insert into meal (title, description, location, meal_time, max_reservations, price, created_date) values 
+('Wine - Magnotta', 'Belpaese,sapien cum sociis natoque penatibus et magnis', '457 Cody Point', '2020/07/14 18:30:00', 10, 45.99, '2020/07/14');
```
Comment (flemeur) [inline databases/week3/meal_sharing_db.sql:37]: It works in MySQL to use any delimiter character in a date/datetime, but the standard way of doing it is using dashes and colons, like this: `2020-07-14 18:30:00`

https://dev.mysql.com/doc/refman/8.0/en/date-and-time-literals.html

File: databases/week3/meal_sharing_db.sql [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+-- Get meals that has a price smaller than a specific price fx 90
+SELECT * FROM meal
+WHERE price < 90;
+-- Get meals that still has available reservations
+
+SELECT m.id, m.title, m.description, m.location, m.price
+FROM meal m
+JOIN reservation r ON m.id = r.meal_id
+WHERE m.meal_time >= CURRENT_TIMESTAMP AND m.max_reservations >= r.number_of_guests
```
Comment (flemeur) [inline databases/week3/meal_sharing_db.sql:136]: Great that you included a check for the meal_time as well as the available reservations :ok_hand: 

One thing to think about is whether the checks should be `>=` or `>`. In terms of available reservations you can't really reserve a spot for a meal where the maximum has been reached.

File: databases/week3/meal_sharing_db.sql [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+-- Get only specific number of meals fx return only 5 meals
+SELECT * FROM meal
+LIMIT 5;
+
+-- Get the meals that have good reviews
+SELECT m.id, m.title, m.description, m.price, r.stars, r.description FROM meal m
+INNER JOIN review r ON m.id = r.meal_id
+WHERE r.stars >= 4
+GROUP BY m.id
+ORDER BY r.stars DESC;
```
Comment (flemeur) [inline databases/week3/meal_sharing_db.sql:156]: This one gives me this error: `Error in query (1055): Expression #5 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'Meal_sharing.r.stars' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by`

Can you figure out what is causing that?

File: databases/week3/meal_sharing_db.sql [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+-- Get reservations for a specific meal sorted by created_date
+SELECT * FROM reservation
+WHERE meal_id = 5
+ORDER BY created_date DESC;
+
+-- Sort all meals by average number of stars in the reviews
+SELECT m.id, m.title, m.description, m.price, r.stars FROM meal m
+INNER JOIN review r ON m.id = r.meal_id
+GROUP BY m.id
+ORDER BY AVG(r.stars);
```
Comment (flemeur) [inline databases/week3/meal_sharing_db.sql:167]: This one gives me the same error: `Error in query (1055): Expression #5 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'Meal_sharing.r.stars' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by`

As a hint i can say that you are very close to the correct query

File: databases/week3/meal_sharing_db.sql (hunk 2 above)
Comment (panivita) [reply databases/week3/meal_sharing_db.sql:37]: I used https://www.mockaroo.com/ for schema creation as you suggested.

File: databases/week3/meal_sharing_db.sql (hunk 4 above)
Comment (panivita) [reply databases/week3/meal_sharing_db.sql:156]: Why I don't have this error? It is working on my machine

File: databases/week3/meal_sharing_db.sql (hunk 5 above)
Comment (panivita) [reply databases/week3/meal_sharing_db.sql:167]: Why I don't have this error? It is working on my machine

File: databases/week3/meal_sharing_db.sql (hunk 1 above)
Comment (panivita) [reply databases/week3/meal_sharing_db.sql:14]: Why I don't have this error? It is working on my machine

Comment (panivita) [reply databases/week3/meal_sharing_db.sql:14]: It is my version MySQL 

innodb_version | 8.0.20
-- | --
original_server_version | 999999
protocol_version | 10
slave_type_conversions |  
tls_version | TLSv1,TLSv1.1,TLSv1.2,TLSv1.3
version | 8.0.20
version_comment | MySQL Community Server - GPL
version_compile_machine | x86_64
version_compile_os | Win64
version_compile_zlib | 1.2.11

File: databases/week3/meal_sharing_db.sql (hunk 5 above)
Comment (panivita) [reply databases/week3/meal_sharing_db.sql:167]: When I delete group by                                                                                                      
 SELECT m.id, m.title, m.description, r.stars, r.description FROM meal m
INNER JOIN review r ON m.id = r.meal_id
ORDER BY AVG(r.stars);
I have error  SELECT m.id, m.title, m.description, r.stars, r.description FROM meal m INNER JOIN review r ON m.id = r.meal_id ORDER BY AVG(r.stars) LIMIT 0, 1000	Error Code: 3029. Expression #1 of ORDER BY contains aggregate function and applies to the result of a non-aggregated query	0.000 sec
But in this case I don't have error
SELECT m.id, m.title, m.description, r.stars, r.description FROM meal m
INNER JOIN review r ON m.id = r.meal_id
GROUP BY m.id
ORDER BY AVG(r.stars);
01:17:24	SELECT m.id, m.title, m.description, r.stars, r.description FROM meal m INNER JOIN review r ON m.id = r.meal_id GROUP BY m.id ORDER BY AVG(r.stars) LIMIT 0, 1000	4 row(s) returned	0.000 sec / 0.000 sec

Comment (panivita) [reply databases/week3/meal_sharing_db.sql:167]: I find this https://stackoverflow.com/questions/5247347/trying-to-use-order-by-with-average-values

File: databases/week3/meal_sharing_db.sql (hunk 4 above)
Comment (panivita) [reply databases/week3/meal_sharing_db.sql:156]: 01:19:50	SELECT m.id, m.title, m.description, r.stars, r.description FROM meal m INNER JOIN review r ON m.id = r.meal_id WHERE r.stars >= 4 GROUP BY m.id ORDER BY r.stars DESC LIMIT 0, 1000	3 row(s) returned	0.000 sec / 0.000 sec

**Pull request merged**